src/components/shapes/draw.js

```diff
@@ -91,10 +91,12 @@ function drawOne(gd, index) {
     // TODO: use d3 idioms instead of deleting and redrawing every time
     if(!options._input || options.visible !== true) return;
 
-    if(options.layer !== 'below') {
+    if(options.layer === 'above') {
         drawShape(gd._fullLayout._shapeUpperLayer);
     } else if(options.xref === 'paper' || options.yref === 'paper') {
         drawShape(gd._fullLayout._shapeLowerLayer);
+    } else if(options.layer === 'between') {
```

Comment:
Pretty sure the order of this case should be swapped with the one before it. If `layer` is explicitly set to `'between'` we don't want it to be drawn on the lower layer just because `xref` or `yref` is set to `'paper'`.

Maybe add this to the test json as well.

Reply:
Problem is, the "between" layer is added in the cartesian layer in order to appear over the gridlines which are also in this layer.
Without looking deep into it, it seems that a paper ref inside this context is not possible.